tests/xdsl_opt/test_xdsl_opt.py

```diff
@@ -156,7 +157,7 @@ def test_diagnostic_exception():
 
     opt = xDSLOptMain(args=[filename_in])
 
-    with pytest.raises(DiagnosticException):
+    with pytest.raises(ParseError):
```

Comment:
This change is needed, as the Parser does now the type-checking and now throws an error: 
> operand is used with type f32, but has been previously used or defined with type i32

Is there a specific reason that we do type-checking in the parser. I feel this is somehow the wrong place to do type checking as it e.g. prevents further type errors at other locations to be emitted.

Reply:
Let me check if we should instead change the program that is being parsed there.

The reason this is checked at parse time and not after, is because the program cannot even be constructed.
If you refer to a value that is expected to have type `i32`, and then you parse that value definition which has a different type, let's say `i64`, then you have a fundamental inconsistency in your program.
It's not even that it is ill-typed, it is that a value has two different types, and thus we should signal it directly.

Reply:
I am confused. In which way would a type error be `fundamentally inconsistent`?

Reply:
What I mean is that you cannot even create an IR that correspond to that program, because you have a value that has two different types.

Also, note that this is different than having an operation not verifying. This is handled later.

Reply:
Also, this is exactly how it is done in MLIR, just in case.